build.gradle.kts

```
@@ -0,0 +1,232 @@
+
+    testImplementation("junit:junit:4.12")
+    testImplementation("org.junit.jupiter:junit-jupiter-api:5.5.2")
+    testRuntimeOnly("org.junit.vintage:junit-vintage-engine:5.5.2")
+
+    testImplementation("org.hamcrest:hamcrest:2.2")
+}
+
+sourceSets {
+    create("functionalTest") {
+        java.srcDir(file("src/functionalTest/java"))
+        compileClasspath += sourceSets["main"].output
+        runtimeClasspath += sourceSets["main"].output
+    }
+    create("crossScalaVersionTest") {
+        java.srcDir(file("src/crossScalaVersionTest/java"))
+        compileClasspath += sourceSets["main"].output + sourceSets["functionalTest"].output
+        runtimeClasspath += sourceSets["main"].output + sourceSets["functionalTest"].output
+    }
+}
+
+configurations {
+    named("functionalTestImplementation") {
+        extendsFrom(configurations["testImplementation"])
+    }
+    named("functionalTestRuntimeOnly") {
+        extendsFrom(configurations["testRuntimeOnly"])
+    }
+
+    named("crossScalaVersionTestImplementation") {
+        extendsFrom(configurations["testImplementation"])
+    }
+    named("crossScalaVersionTestRuntimeOnly") {
+        extendsFrom(configurations["testRuntimeOnly"])
+    }
+}
+
+val crossScalaVersionTest by tasks.registering(Test::class) {
+    description = "Runs the cross scala version functional test."
+    group = "verification"
+    testClassesDirs = sourceSets["crossScalaVersionTest"].output
+    classpath = sourceSets["crossScalaVersionTest"].runtimeClasspath
+    setForkEvery(1) // crucial to run every test in its own JVM
+
+    testLogging {
+        events("passed", "failed", "skipped")
+        showStandardStreams = System.getenv("CI") == "true"
+    }
+
+    mustRunAfter(tasks["test"])
+}
+tasks["check"].dependsOn(crossScalaVersionTest)
+
+val functionalTest by tasks.registering(Test::class) {
+    description = "Runs the functional tests."
+    group = "verification"
+    testClassesDirs = sourceSets["functionalTest"].output
+    classpath = sourceSets["functionalTest"].runtimeClasspath
+
+    testLogging {
+        events("passed", "failed", "skipped")
+        showStandardStreams = System.getenv("CI") == "true"
+    }
+
+    systemProperty("failOnWarning", project.hasProperty("failOnWarning"))
+
+    mustRunAfter(crossScalaVersionTest)
+}
+tasks["check"].dependsOn(functionalTest)
+
+gradlePlugin {
+    testSourceSets(sourceSets["functionalTest"], sourceSets["crossScalaVersionTest"])
+}
+
+val kotlindocJar by tasks.registering(Jar::class) {
+    from(tasks.dokkaHtml.get().outputDirectory)
+    archiveClassifier.set("kotlindoc")
+    dependsOn(tasks.dokkaHtml)
+}
+
+val sourcesJar by tasks.registering(Jar::class) {
+    from(sourceSets["main"].allSource)
+    archiveClassifier.set("sources")
+}
+
+fun propOrDefault(property: String): String {
+    if (project.hasProperty(property)) {
+        return project.property(property).toString()
+    } else {
+        return ""
+    }
+}
```

This can be more Kotlin-idiomatically written like this:
```suggestion
fun Project.propOrDefault(property: String) =
  if (hasProperty(property)) property(property).toString() else ""
```
It is also more generic now since it can be applied to any `Project` instance, not just the current project instance.